src/finn/custom_op/fpgadataflow/matrixvectoractivation.py

```diff
@@ -971,3 +956,145 @@ def code_generation_ipi(self):
+
+        tracker = 0
+        maximum = numVectors * BURST_SIZE
+
+        if numVectors > 1:
+            for i in range(2):
+                txns.append(counter)
+                counter += 1
+                cycles += 1
+                tracker += 1
+
+        for k in range(numVectors):
+            for j in range(BURST_SIZE):
+                if tracker < maximum:
+                    txns.append(counter)
+                    counter += 1
+                    cycles += 1
+                    tracker += 1
+
+            for i in range(BURST_COUNT - 1):
+                for j in range(BURST_SIZE):
+                    txns.append(counter)
+                    cycles += 1
+
+        return txns, cycles, counter
+
+    def characteristic_fx_output(self, txns, cycles, counter, kwargs):
+        (MW, MH, SIMD, PE, BURST_COUNT, BURST_SIZE, numVectors) = kwargs
+
+        windup_clocks = 3
+
+        for i in range(0, windup_clocks):
+            txns.append(counter)
+            cycles += 1
+
+        for k in range(numVectors):
+            for i in range(BURST_COUNT):
+                for j in range(BURST_SIZE):
+                    txns.append(counter)
+                    cycles += 1
+                counter += 1
+
+        return txns, cycles, counter
+
+    def derive_characteristic_fxns(self, period):
+        n_inps = np.prod(self.get_folded_input_shape()[:-1])
+        io_dict = {
+            "inputs": {
+                "in0": [0 for i in range(n_inps)],
+            },
+            "outputs": {"out": []},
+        }
+
+        mem_mode = self.get_nodeattr("mem_mode")
+        if mem_mode in ["internal_decoupled", "external"]:
+            n_weight_inps = self.calc_wmem()
+            num_w_reps = np.prod(self.get_nodeattr("numInputVectors"))
+            io_dict["inputs"]["weights"] = [0 for i in range(num_w_reps * n_weight_inps)]
+
+        ignore = self.get_nodeattr("ipgen_ignore")
+        if ignore is False:  # this node is being derived using RTLSIM
+            # RTL-based flow
+            super().derive_characteristic_fxns(period, override_rtlsim_dict=io_dict)
+            return
+
+        # Analytical flow
+
+        txns_in = {key: [] for (key, value) in io_dict["inputs"].items() if "in" in key}
+        txns_out = {key: [] for (key, value) in io_dict["outputs"].items() if "out" in key}
+
+        all_txns_in = np.empty((len(txns_in.keys()), 2 * period), dtype=np.int32)
+        all_txns_out = np.empty((len(txns_out.keys()), 2 * period), dtype=np.int32)
+
+        self.set_nodeattr("io_chrc_period", period)
+
+        txn_in = []
+        txn_out = []
+
+        # INPUT
+
+        counter = 0
+        padding = 0
+
+        kwargs = self.prepare_kwargs_for_characteristic_fx()
+
+        # first period
+        cycles = 0
+        txn_in, cycles, counter = self.characteristic_fx_input(txn_in, cycles, counter, kwargs)
+
+        txn_in += [counter] * (period - cycles)
+        padding += period * -cycles
```

Comment:
Is this a typo, i.e. `padding += (period * 1) - cycles`?

Reply:
Yes, will fix in the next push sometime this or next week.